### ISSUE TITLE
Using number escapes (binary, octal, etc) with variables

### ISSUE THREAD

**Issue by matthewwiese** [arclanguage/anarki#98]
I'm looking to implement simple bitwise left and right shifts in Arc for a personal project.

My first dilemma is simply properly referring to variables using the `#b` notation:
```
arc> (= q 10)
10
arc> #bq
```
throws an error of `read: bad number: q` and so does `#b\q`, whereas:
```
arc> #b\10
2
```
evaluates just fine.

I dug through the documentation and I'm unable to find any mention of the proper way to do this.

**Comment by akkartik** [comment]
I haven't looked at this particular feature, but I'd be very surprised if `#` forms were anything but literals.

`q` has the value 'ten' regardless of base. It would be weird for it to remember the sequence of digits that that value arose from.

**Comment by matthewwiese** [comment]
What would be the most efficient method of storing a binary number in Arc?

My idea above was a hack-y approach of just using ones and zeroes in a typical decimal number. Arc believes `q` has the value of _ten_, but I want it to interpret _10_ as the binary representation of _2_. Such that the escape sequence `#b` is "always" there.

**Comment by akkartik** [comment]
Can you back up and describe the larger goal that requires binary numbers?

Looking at the history of this issue, I don't think you strictly need binary representation to do bit shifts.

For example, this seems like a reasonable session in Arc:

```lisp
arc> (lshift 3 2)  ; shift 0x000011 left by 2 bits
12                 ;       0x001100
```

This should work regardless of base. And then you just need to provide primitives for displaying numbers in different bases. Independent of shift operations.

> What would be the most efficient method of storing a binary number in Arc?

I don't follow that question. Computers store just numbers, independent of base. Usually considerations of the base/radix enter the picture only at time of display. Or when parsing numbers from strings

**Comment by matthewwiese** [comment]
The purpose is to implement xorshifting for a pseudorandom number generator.

Your example of `lshift` is exactly what I'm looking to do. Problem is, I'm not entirely sure how to go about it in Arc given the primitives.

> I don't follow that question. Computers store just numbers, independent of base. Usually considerations of the base/radix enter the picture only at time of display. Or when parsing numbers from strings

Yeah that was a bit simpleminded of me. Let me try again using Common Lisp as an example. CL has `ash` for arithmetic shifting and `logand` for logical conjunction. Using these two functions, one can implement your `lshift` example (and `rshift`!) themselves.

However, there is no equivalent to `ash` in Arc. Which leaves me at a loss for how to implement it myself.

**Comment by akkartik** [comment]
Ah, I see.

Yeah, it's sadly pretty common for Arc to be missing primitives. When that happens it's useful to see if the underlying Racket has the primitive you want. (It's nice and comprehensive, so the answer is usually "yes".)

In this case, googling for 'racket bit shift' points me at https://docs.racket-lang.org/reference/generic-numbers.html, which may be what you're looking for. Integrating these Racket primitives with Arc may require using `xdef` ([example](https://github.com/arclanguage/anarki/blob/75b264fe6427b70091f9715b367139bcbde9f569/ac.rkt#L756)) or `$` ([example](https://github.com/arclanguage/anarki/blob/75b264fe6427b70091f9715b367139bcbde9f569/lib/re.arc)).

Things changed a bit recently with how Arc interacts with Racket. I'm not familiar with those, so I'll tag @rocketnia in case you end up having to deal with racket packages :)

**Comment by hjek** [comment]
On Tue, Mar 06, 2018 at 11:54:08AM -0800, matt wrote:
> Your example of `lshift` is exactly what I'm looking to do. Problem is, I'm not entirely sure how to go about it in Arc given the primitives.

You could just use the Racket primitive, no?


```
arc> #b1000
8
arc> ($.arithmetic-shift #b1000 2)
32
arc> ($.arithmetic-shift #b1000 1)
16
arc> ($.arithmetic-shift #b1000 -2)
2
```

If you need to print binary numbers, do this:
```
arc> ($.format "~b" 1024)
"10000000000"
arc> ($.format "~b" 8)
"1000"
```

In general you can use any Racket function with the `$` escape.
